controllers/storageclient_controller.go

```diff
@@ -0,0 +1,307 @@
+		instance.ObjectMeta.Finalizers = append(instance.ObjectMeta.Finalizers, storageClientFinalizer)
+		if err := s.Client.Update(s.ctx, instance); err != nil {
+			s.Log.Info("Failed to update StorageClient with finalizer.", "StorageClient", klog.KRef(instance.Namespace, instance.Name))
+			return reconcile.Result{}, fmt.Errorf("failed to update StorageClient with finalizer: %v", err)
+		}
+	}
+
+	if instance.Status.ConsumerID == "" {
+		return s.onboardConsumer(instance, externalClusterClient)
+	} else if instance.Status.Phase == v1alpha1.StorageClientOnboarding {
+		return s.acknowledgeOnboarding(instance, externalClusterClient)
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (s *StorageClientReconciler) deletionPhase(instance *v1alpha1.StorageClient, externalClusterClient *providerClient.OCSProviderClient) (ctrl.Result, error) {
+	// TODO Need to take care of deleting the SCC created for this
+	// storageClient and also the default SCC created for this storageClient
+	if contains(instance.GetFinalizers(), storageClientFinalizer) {
+		instance.Status.Phase = v1alpha1.StorageClientOffboarding
+		if res, err := s.offboardConsumer(instance, externalClusterClient); err != nil {
+			s.Log.Error(err, "Offboarding in progress.")
+		} else if !res.IsZero() {
+			// result is not empty
+			return res, nil
+		}
+		s.Log.Info("removing finalizer from StorageClient.", "StorageClient", klog.KRef(instance.Namespace, instance.Name))
+		// Once all finalizers have been removed, the object will be deleted
+		instance.ObjectMeta.Finalizers = remove(instance.ObjectMeta.Finalizers, storageClientFinalizer)
+		if err := s.Client.Update(s.ctx, instance); err != nil {
+			s.Log.Info("Failed to remove finalizer from StorageClient", "StorageClient", klog.KRef(instance.Namespace, instance.Name))
+			return reconcile.Result{}, fmt.Errorf("failed to remove finalizer from StorageClient: %v", err)
+		}
+	}
+	s.Log.Info("StorageClient is offboarded", "StorageClient", klog.KRef(instance.Namespace, instance.Name))
+	return reconcile.Result{}, nil
+}
+
+// newExternalClusterClient returns the *providerClient.OCSProviderClient
+func (s *StorageClientReconciler) newExternalClusterClient(instance *v1alpha1.StorageClient) (*providerClient.OCSProviderClient, error) {
+
+	ocsProviderClient, err := providerClient.NewProviderClient(
+		s.ctx, instance.Spec.StorageProviderEndpoint, time.Second*10)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create a new provider client: %v", err)
+	}
+
+	return ocsProviderClient, nil
+}
+
+// onboardConsumer makes an API call to the external storage provider cluster for onboarding
+func (s *StorageClientReconciler) onboardConsumer(instance *v1alpha1.StorageClient, externalClusterClient *providerClient.OCSProviderClient) (reconcile.Result, error) {
+
+	// TODO Need to find a way to this one as its OCP dependent
```

Comment:
I have no idea what this is saying, I think there's a verb missing?

Reply:
Reworded it.